### PR TITLE
fix(recsys): resolve KeyError 'post_id' in rec_sys_personalized_with_trace

### LINKED ISSUE #198: [BUG] rec_sys_personalized_with_trace() access an invalid key
### Required prerequisites

- [x] I have read the documentation <https://docs.oasis.camel-ai.org/>.
- [x] I have searched the [Issue Tracker](https://github.com/camel-ai/oasis/issues) and [Discussions](https://github.com/camel-ai/oasis/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [ ] Consider asking first in a [Discussion](https://github.com/camel-ai/oasis/discussions/new).

### What version of camel-oasis are you using?

0.2.5

### System information

installed via uv

3.11.14 (main, Dec  9 2025, 19:02:23) [Clang 21.1.4 ] linux

### Problem description

in oasis/social_platform/recsys.py
in the function rec_sys_personalized_with_trace()

Image

'post_id' is not a valid key for trace -> raise an error

### Reproducible example code

slightly modified version of
examples/experiment/twitter_simulation_1M_agents/twitter_simulation_1m.py

when calling
await platform.update_rec_table()

### Traceback

```pytb
Traceback (most recent call last):
  File "/home/noce/.local/share/uv/python/cpython-3.11.14-linux-x86_64-gnu/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/home/noce/.local/share/uv/python/cpython-3.11.14-linux-x86_64-gnu/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/noce/.local/share/uv/python/cpython-3.11.14-linux-x86_64-gnu/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/noce/Documents/TG/backend/src/services/simulation_service.py", line 234, in run
    await self.platform.update_rec_table()
  File "/home/noce/Documents/TG/backend/.venv/lib/python3.11/site-packages/oasis/social_platform/platform.py", line 340, in update_rec_table
    new_rec_matrix = rec_sys_personalized_with_trace(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/noce/Documents/TG/backend/.venv/lib/python3.11/site-packages/oasis/social_platform/recsys.py", line 784, in rec_sys_personalized_with_trace
    swap_free_ids = [
                    ^
  File "/home/noce/Documents/TG/backend/.venv/lib/python3.11/site-packages/oasis/social_platform/recsys.py", line 786, in <listcomp>
    if post_id not in rec_post_ids and post_id not in [
                                                      ^
  File "/home/noce/Documents/TG/backend/.venv/lib/python3.11/site-packages/oasis/social_platform/recsys.py", line 787, in <listcomp>
    trace['post_id']
    ~~~~~^^^^^^^^^^^
KeyError: 'post_id'
```

### Expected behavior

Not raising an error

### Additional context

Image

trace only have these keys

### PR DESCRIPTION
**Pull request by Ghraven**
Fixes #198

## Problem

Calling `platform.update_rec_table()` with a non-zero `swap_rate` raises `KeyError: 'post_id'`:

```
File "oasis/social_platform/recsys.py", line 787
    trace['post_id']
KeyError: 'post_id'
```

## Root Cause

The `trace` table schema stores interaction data in an `info TEXT` column as a JSON string — there is no top-level `post_id` column (see `schema/trace.sql`). The code at line 787 accessed `trace['post_id']` directly, which does not exist.

The rest of the codebase correctly reads post IDs from the `info` field via `literal_eval(trace['info'])["post_id"]` (e.g. `get_like_post_id()` at L371).

## Fix

Replace the direct `trace['post_id']` access with `literal_eval(trace['info']).get('post_id')`, consistent with the existing pattern. Use `.get()` instead of `[]` to safely skip trace entries for actions that do not produce a `post_id` (e.g. follow, retweet). Pre-compute the set of traced post IDs outside the list comprehension to avoid redundant work per post.

## Before / After

```python
# Before (raises KeyError)
swap_free_ids = [
    post_id for post_id in post_ids
    if post_id not in rec_post_ids and post_id not in [
        trace['post_id']
        for trace in trace_table if trace['user_id']
    ]
]

# After
traced_post_ids = {
    literal_eval(trace['info']).get('post_id')
    for trace in trace_table
    if trace.get('user_id')
}
swap_free_ids = [
    post_id for post_id in post_ids
    if post_id not in rec_post_ids
    and post_id not in traced_post_ids
]
```

### CODE REVIEW

File: oasis/social_platform/recsys.py [hunk 1]
```diff
@@ -781,12 +781,15 @@ def rec_sys_personalized_with_trace(
...
+                }
                 swap_free_ids = [
                     post_id for post_id in post_ids
-                    if post_id not in rec_post_ids and post_id not in [
-                        trace['post_id']
-                        for trace in trace_table if trace['user_id']
-                    ]
+                    if post_id not in rec_post_ids
+                    and post_id not in traced_post_ids
                 ]
```
[the remark this answers is not in the document]
Comment (echo-yiyiyi) [reply oasis/social_platform/recsys.py:796]: @copilot apply changes based on [this feedback](https://github.com/camel-ai/oasis/pull/218#discussion_r3207382284)

File: oasis/social_platform/recsys.py [hunk 2]
```diff
@@ -781,12 +781,15 @@ def rec_sys_personalized_with_trace(
 
             if swap_rate > 0:
                 # swap the recommended posts with random posts
+                traced_post_ids = {
+                    literal_eval(trace['info']).get('post_id')
+                    for trace in trace_table
+                    if trace.get('user_id')
+                }
```
[the remark this answers is not in the document]
Comment (echo-yiyiyi) [reply oasis/social_platform/recsys.py]: @copilot apply changes based on [this feedback](https://github.com/camel-ai/oasis/pull/218#discussion_r3207382265)

File: oasis/social_platform/recsys.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (echo-yiyiyi) [reply oasis/social_platform/recsys.py:796]: @copilot give code suggestion based on [this feedback](https://github.com/camel-ai/oasis/pull/218#discussion_r3207382284)

File: oasis/social_platform/recsys.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (echo-yiyiyi) [reply oasis/social_platform/recsys.py]: @copilot give code suggestion based on https://github.com/camel-ai/oasis/pull/218#discussion_r3207382265

**Review by echo-yiyiyi — COMMENTED** [review]
Thanks @Ghraven very much for your PR. Since copilot is not available in your forked repo. Could you modify the code based on the advice of the copilot here. I read them and agree with it, especially for adding the test.

**Comment by Ghraven** [comment]
Thanks for the thorough review, @echo-yiyiyi and @Copilot! Really appreciate the detailed feedback.

All three points are valid — I'll address them:

1. **`user_id` falsy check** — will update the filter to `trace.get('user_id') is not None` to safely handle `user_id=0` edge cases.
2. **`traced_post_ids` recomputation** — will move it outside the per-user loop since it only depends on `trace_table`. Good catch on the unnecessary overhead.
3. **Missing test** — will add a unit test for `rec_sys_personalized_with_trace()` covering `swap_rate > 0` with trace rows that lack a `post_id` (e.g., follow/unfollow actions).

Will push the updated changes shortly. Thanks again for taking the time!

**Comment by Ghraven** [comment]
Just pushed the remaining fixes based on the review feedback:

**Commit 1** — [`8b1dacd`](https://github.com/Ghraven/oasis/commit/8b1dacdfc4594114388a05d3e7965285afecae7f): moved `traced_post_ids` outside the per-user `for` loop. Since it only depends on `trace_table` (not the current user), it was being redundantly recomputed on every iteration. Now computed once before the loop.

**Commit 2** — [`75e1382`](https://github.com/Ghraven/oasis/commit/75e1382d98cbc1eaba9f21f60d741d284dc66337): added `tests/test_recsys_trace.py` with 5 unit tests covering:
- `user_id=0` not incorrectly excluded (falsy fix)
- follow/unfollow traces with no `post_id` do not raise `KeyError`
- `swap_rate=0` skips trace computation entirely
- empty `trace_table` with `swap_rate > 0`
- `traced_post_ids` consistent across users

Thanks again for the detailed review — really helped tighten this up. Let me know if anything else needs adjusting!

**Comment by Ghraven** [comment]
Hi! Just a gentle ping on this bug fix. Happy to add tests or revise the approach if that would help move it forward. Thanks for your time!
